### PR TITLE
buck: init at v2017.05.31.01

### PR DESCRIPTION
**Pull request by jgertm**
###### Motivation for this change

This PR introduces the `buck` build tool, developed at Facebook. It contains merely the static .pex file generated by `buck build buck`, not the recommended mutable directory that buck uses to satisfy the version specified by a project's `.buckversion` file. 

Such is life in Nixistan.

###### Things done

- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/tools/build-managers/buck/default.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    repo = "buck";
+    rev = "0b8b3828a11afa79dc128832cb55b106f07e48aa";
+    sha256 = "1g3yg8qq91cdhsq7zmir7wxw3767l120f5zhq969gppdw9apqy0s";
+  };
+
+  patches = [ ./pex-mtime.patch ];
+
+  postPatch = ''
+    for f in $(grep -l -r '/bin/bash'); do
+      substituteInPlace "$f" --replace '/bin/bash' '${bash}/bin/bash'
```
Comment (Mic92) [inline pkgs/development/tools/build-managers/buck/default.nix]: Do you only replace shebangs here? If so, add bash as input and use `patchShebangs .`.

Comment (Mic92) [inline pkgs/development/tools/build-managers/buck/default.nix]: it is not only used as shebang.

File: pkgs/development/tools/build-managers/buck/default.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  buildPhase = ''
+    ant
+    ./bin/buck build buck
+  '';
+
+  installPhase = ''
+    install -D -m755 buck-out/gen/programs/buck.pex $out/bin/buck
+    wrapProgram $out/bin/buck \
+      --prefix PYTHONPATH : $PYTHONPATH \
+      --prefix PATH : "${stdenv.lib.makeBinPath [jre watchman]}"
```
Comment (Mic92) [inline pkgs/development/tools/build-managers/buck/default.nix]: It requires java at runtime. Please check that wheter `jre` or `jdk` is required here.

**Comment by Mic92** [comment]
Did you also test to build projects in some languages? Java, C++, Android etc?

File: pkgs/development/tools/build-managers/buck/default.nix (hunk 2 above)
Comment (Mic92) [inline pkgs/development/tools/build-managers/buck/default.nix]: `jre_headless` might be also an option to reduce the closure size.

File: pkgs/development/tools/build-managers/buck/default.nix [hunk 3]
```diff
@@ -1,6 +1,8 @@
-  { stdenv, pkgs, fetchFromGitHub, fetchgit, jdk, ant, python2, watchman, unzip, bash }:
+{ stdenv, fetchFromGitHub, jdk, ant, python2, python2Packages, watchman, unzip, bash, jre, makeWrapper }:
 
-stdenv.mkDerivation rec {
+
+let
```
Comment (jgertm) [inline pkgs/development/tools/build-managers/buck/default.nix]: This seems to have no purpose.

**Comment by jgertm** [comment]
It builds a really big Clojure codebase, so I think it checks out on that front.

File: pkgs/development/tools/build-managers/buck/default.nix (hunk 2 above)
Comment (jgertm) [inline pkgs/development/tools/build-managers/buck/default.nix]: `buck` is primarily used as a Java build tool, so I think it's okay to include JDK.

**Comment by Mic92** [comment]
Thanks!

**Pull request merged**